sql: remove unused backwards compatibility type mutations

paulniziolek opened the pull request:
#### sql: remove unused backwards compatibility type mutations

Removed upgradeType() and downgradeType() backwards compatibility features
(that only applied to clusters before v19.1) that were unused and bug-prone
when implementing some newer types.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148283)
<!-- Reviewable:end -->

paulniziolek commented on pkg/sql/types/types.go:
```diff
@@ -2682,79 +2416,7 @@ func (t *T) MarshalToSizedBuffer(data []byte) (int, error) {
-		}
-	}
-
-	// Map empty locale to nil.
-	if t.InternalType.Locale != nil && len(*t.InternalType.Locale) == 0 {
-		t.InternalType.Locale = nil
-	}
```
We probably don't need to do this anymore?

paulniziolek committed:
sql: remove unused backwards compatibility type mutations

Removed upgradeType() and downgradeType() backwards compatibility features
(that only applied to clusters before v19.1) that were unused and bug-prone
when implementing some newer types.

Epic: None
Release note: None
